cray_product_catalog/util/merge_dict.py

```diff
@@ -0,0 +1,156 @@
+    """Return true if the given dict has no list or dictionary values.
+
+    Helper for merge_dict.
+
+    Args:
+        dict_to_check (dict): The dict to check for list/dictionary values.
+
+    Returns:
+        bool: True if the dict has no dict or list values.
+    """
+    return not any(isinstance(value, dict) or isinstance(value, list)
+                   for value in dict_to_check.values())
+
+
+def _values_are_dicts(*values):
+    """Return true if all given values are dict type.
+
+    Helper for merge_dict.
+
+    Args:
+        *values: A list of values to check.
+
+    Returns:
+        bool: True if the values are dicts.
+    """
+    return all(isinstance(v, dict) for v in values)
+
+
+def _values_are_lists(*values):
+    """Return true if all values are list type.
+
+    Helper for merge_dict.
+
+    Args:
+        *values: A list of values to check.
+
+    Returns:
+        bool: True if the values are lists.
+    """
+    return all(isinstance(v, list) for v in values)
+
+
+def _values_are_different_types(first_value, second_value):
+    """Return true if two values are not the same type.
+
+    Helper for merge_dict.
+
+    Args:
+        first_value: the value to compare with second_value.
+        second_value: the value to compare with first_value.
+
+    Returns:
+        bool: True if the two values are not the same type.
+    """
+    return not (
+        isinstance(first_value, type(second_value)) or
+        isinstance(second_value, type(first_value))
+    )
+
+
+def _merge_input_with_existing(input_key, input_value, dict_to_update):
+    """Merge the given input with an existing dictionary.
+
+    Helper for merge_dict.
+
+    Args:
+        input_key: The key under which to insert the new data.
+        input_value: The new value to insert.
+        dict_to_update: The dictionary in which to insert the new data.
+
+    Returns:
+        None. Modifies dict_to_update in place.
+
+    Raises:
+        ValueError: if two values contain conflicting types that can't
+            be merged (e.g. merging a string with a list).
+    """
+    # If adding a new key not in the existing dict, just add it.
+    if input_key not in dict_to_update:
+        dict_to_update[input_key] = input_value
+    else:
+        if _values_are_dicts(dict_to_update[input_key], input_value):
+            # Merging two dicts, use merge_dict() recursively.
+            dict_to_update[input_key] = merge_dict(input_value, dict_to_update[input_key])
+        elif _values_are_lists(dict_to_update[input_key], input_value):
+            # Merging two lists, use extend().
+            dict_to_update[input_key].extend(input_value)
+        else:
+            if _values_are_different_types(input_value, dict_to_update[input_key]):
+                raise ValueError(
+                    f'Cannot merge {input_value} (type {type(input_value)}) '
+                    f'with {dict_to_update[input_key]} (type {type(dict_to_update[input_key])})'
```

Comment:
nit: `TypeError` might be more appropriate here

Reply:
Thanks, fixed in 24778dd5947ad8f48dbadf2ded6d403a2d3ca6b6